src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -0,0 +1,179 @@
+const READ_BASE_LAYER = 0;
+
+class F extends ValidationTest {
+  createTexture(
+    options: {
+      width?: number;
+      height?: number;
+      arrayLayerCount?: number;
+      mipLevelCount?: number;
+      sampleCount?: number;
+      format?: GPUTextureFormat;
+    } = {}
+  ): GPUTexture {
+    const {
+      width = 32,
+      height = 32,
+      arrayLayerCount = 1,
+      mipLevelCount = 1,
+      sampleCount = 1,
+      format = 'rgba8unorm',
+    } = options;
+
+    return this.device.createTexture({
+      size: { width, height, depth: arrayLayerCount },
+      mipLevelCount,
+      sampleCount,
+      dimension: '2d',
+      format,
+      usage: GPUTextureUsage.OUTPUT_ATTACHMENT | GPUTextureUsage.SAMPLED,
+    });
+  }
+
+  getDescriptor(
+    options: {
+      format?: GPUTextureFormat;
+      dimension?: GPUTextureViewDimension;
+      aspect?: GPUTextureAspect;
+      baseMipLevel?: number;
+      mipLevelCount?: number;
+      baseArrayLayer?: number;
+      arrayLayerCount?: number;
+    } = {}
+  ): GPUTextureViewDescriptor {
+    const {
+      format = 'rgba8unorm',
+      dimension = '2d',
+      aspect = 'all',
+      baseMipLevel = 0,
+      mipLevelCount = 1,
+      baseArrayLayer = 0,
+      arrayLayerCount = 1,
+    } = options;
+    return {
+      format,
+      dimension,
+      aspect,
+      baseMipLevel,
+      mipLevelCount,
+      baseArrayLayer,
+      arrayLayerCount,
+    };
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('readwrite_on_different_types_of_subresources')
+  .params([
+    // read and write usages are binding to the same texture subresource.
+    {
+      format: 'rgba8unorm' as const,
+      writeBaseLevel: READ_BASE_LEVEL,
+      writeBaseLayer: READ_BASE_LAYER,
+      writeAspect: 'all' as const,
+      success: false,
+    },
+
+    // read and write usages are binding to different mip levels of the same texture.
+    {
+      format: 'rgba8unorm' as const,
+      writeBaseLevel: READ_BASE_LEVEL + 1,
+      writeBaseLayer: READ_BASE_LAYER,
+      writeAspect: 'all' as const,
+      success: true,
+    },
+
+    // read and write usages are binding to different array layers of the same texture.
+    {
+      format: 'rgba8unorm' as const,
+      writeBaseLevel: READ_BASE_LEVEL,
+      writeBaseLayer: READ_BASE_LAYER + 1,
+      writeAspect: 'all' as const,
+      success: true,
+    },
+
+    // read and write usages are binding to different aspects of the same texture.
+    {
+      format: 'depth24plus-stencil8' as const,
+      writeBaseLevel: READ_BASE_LEVEL,
+      writeBaseLayer: READ_BASE_LAYER,
+      writeAspect: 'stencil-only' as const,
+      success: true,
+    },
+  ])
+  .fn(async t => {
+    const { format, writeBaseLevel, writeBaseLayer, writeAspect, success } = t.params;
+
+    const texture = t.createTexture({ arrayLayerCount: 2, mipLevelCount: 6, format });
+    const readAspect = format === 'depth24plus-stencil8' ? 'depth-only' : 'all';
```

Comment:
nit: Pass this in through .params() as a private parameter (`_readAspect`) so you don't have to encode this switch in the test code. You can have it default to 'all' so it's only specified for the one test case that needs it.

Reply:
Not needed after separating tests.